[Documentation] Add documentation for how to add a new biome (#918)
Despair-Games/poketernity: src/data/biomes/README.md

### How to add a biome

#### Pokemon Pool

> A biome's pokemonPool consists of a mapping of a BiomePoolTier (common, uncommon, rare, super rare, ultra rare, boss, boss rare, boss super rare, boss ultra rare) to time of day (dawn, day, dusk, night all), and finally to the pokemon that appear. The current time of day is combined with the ALL pool to determine what Pokemon can appear. The rarirty is rolled first, and if there are no available selections in that tier, it is automatically downgraded according to the above order in BiomePoolTier.

Reviewer: ```suggestion
A biome's pokemonPool consists of a mapping of a BiomePoolTier (common, uncommon, rare, super rare, ultra rare, boss, boss rare, boss super rare, boss ultra rare) to time of day (dawn, day, dusk, night, all), and finally to the pokemon that appear. The current time of day is combined with the ALL pool to determine what Pokemon can appear. The rarity is rolled first, and if there are no available selections in that tier, it is automatically downgraded according to the above order in BiomePoolTier.
```

> A biome's pokemonPool consists of a mapping of a BiomePoolTier (common, uncommon, rare, super rare, ultra rare, boss, boss rare, boss super rare, boss ultra rare) to time of day (dawn, day, dusk, night, all), and finally to the pokemon that appear. The current time of day is combined with the ALL pool to determine what Pokemon can appear. The rarity is rolled first, and if there are no available selections in that tier, it is automatically downgraded according to the above order in BiomePoolTier.

Reviewer 2: ```suggestion
A biome's `pokemonPool` consists of a mapping of a `BiomePoolTier` (common, uncommon, rare, super rare, ultra rare, boss, boss rare, boss super rare, boss ultra rare) to time of day (dawn, day, dusk, night, all), and finally to the pokemon that appear. The current time of day is combined with the ALL pool to determine what Pokemon can appear. The rarity is rolled first, and if there are no available selections in that tier, it is automatically downgraded according to the above order in BiomePoolTier.
```

> Sam's general design philosophy had all the boss tiers contain Pokemon one tier higher than the non boss tier. Boss rare and uncommon would share Pokemon, Boss super rare and rare, Boss ultra rare and super rare, with Boss ultra rare being reserved for the truly rare Pokemon.

Reviewer: ```suggestion
Sam's general design philosophy had all the boss tiers contain Pokemon one tier higher than the non boss tier. `BOSS` would use some Pokemon from `COMMON`, `UNCOMMON`, and `RARE`; `BOSS_RARE` would use some Pokemon from `RARE` and `SUPER_RARE`; `BOSS_SUPER_RARE` would use Pokemon from `ULTRA_RARE`; and `BOSS_ULTRA_RARE` would be reserved for the truly rare Pokemon (usually box legends).
```

Reviewer: <details><summary>Some data from a quick script that I made (includes every biome except for END)</summary>
<p>

![20250408 Boss Distributions](https://github.com/user-attachments/assets/dbadfcef-f997-4448-a978-62ec16526309)

</p>
</details>

Author: Speaking of scripts, I know I removed the pokemon->biome mapping previously, but now I'm wondering if it would be good to have some kind of resource for that in case of a Pokedex where a player would want to look up where Combusken could be found. Of course that would also mean having to simulate evolutions and de evolutions up and down the line now.

Reviewer 2: When we add a pokedex we'd probably add a function that does that at the same time.

For the most part, there is no need to specify evolutions in the Pokemon pool since Pokemon will automatically evolve or de-evolve based on level. The only reason to specify an evolved form is wanting to guarantee that form shows up. (Specifying Gardevoir means Ralts, Kirlia, and Gardevoir can appear, but not Gallade). For Pokemon with specific forms, check the bottom section for how to determine that based on biome.

#### Trainer Pool

> The trainer pool works much in the same way as the Pokemon Pool but does not factor in time of day. Gym leaders are exclusively found in the BOSS tier but there are a lot of unused tiers in the trainer pool. Also like the Pokemon Pool, if there are no available options then the tier is automatically downgraded. If there truly are no trainers aviailable but a trainer needs to spawn, then the default is a breeder.

Reviewer 2: ```suggestion
The trainer pool works much in the same way as the Pokemon Pool but does not factor in time of day. Gym leaders are exclusively found in the `BOSS` tier but there are a lot of unused tiers in the trainer pool. Also like the Pokemon Pool, if there are no available options then the tier is automatically downgraded. If there truly are no trainers aviailable but a trainer needs to spawn, then the default is a breeder.
```

> The trainer chance is a number that represents the demoninator of how likely a trainer battle is. For this reason the chance can only be a fraction like 1/2, 1/3, 1/4, and so on. A trainer chance of 0 means no trainers can spawn.

Reviewer 2: ```suggestion
The trainer chance is a number that represents the demoninator of how likely a trainer battle is. For this reason the chance can only be a fraction like `1/2`, `1/3`, `1/4`, and so on. A trainer chance of `0` means no trainers can spawn.
```

#### Weather Pool and Terrain Pool
These represent the likelihood of weather/terrain being set when the player enters the biome. Keep in mind that the weight of Sun is automatically set to 0 in `arena.setRandomWeather` if the time of day is dusk or night.

### Assets

#### bgm and bgm loop point

> The bgm is the spring representation of the background music of a biome saved as a mp3 like `public/audio/bgm/town.mp3`. Right now all the bgm's are named as the lowercase representation of the biome id, but this can be set to anything.

Reviewer: ```suggestion
All BGM files must be stored in the `public/audio/bgm` directory, then the `bgm` parameter passed into a biome's constructor should match the name of the BGM file. For example, if the file is saved as `public/audio/bgm/town.mp3`, the `bgm` parameter should be `town`. Right now all the bgm's are named as the lowercase representation of the biome ID. This can be set to anything, but make sure to update the corresponding keys in `bgm-name.json` translation files.
```

> The loop point is a number that represents the loop point of a bgm in seconds.

Reviewer: ```suggestion
The loop point of a bgm, measured in seconds, should be set to ensure that the bgm loops/replays smoothly after the bgm file's ending.
```

#### Props and bg

> biomeName_a represents the player's field circle.
> biomeName_b represents the opponent's field circle.
> biomeName_bg represents the background of the biome

Reviewer 2: ```suggestion
`biomeName_a` represents the player's field circle.
`biomeName_b` represents the opponent's field circle.
`biomeName_bg` represents the background of the biome
```

> Props are extra items that show up on the opponent's side of the field. They are labeled as biomeName_b_1 or biomeName_b_2 or biomeName_b_3

Reviewer 2: ```suggestion
Props are extra items that show up on the opponent's side of the field. They are labeled as `biomeName_b_1` or `biomeName_b_2` or `biomeName_b_3`
```

> Power plant and end also have animated biome props that have associated json's

Reviewer 2: ```suggestion
Power Plant and End also have animated biome props that have associated json files.
```

If a biome is NOT an indoor biome, then it is affected by tinting in `field-sprite.onBind`

> ### Other code that needs updating

Reviewer: It turns out a few more biome-dependent things also need to get updated. Here is what I found so far:
- Milcery -> Alcremie evolution
- Dancing Lessons (the Oricorio's form depends on the biome)
- A bunch of code in `mystery-encounters.ts` for which encounters can occur in each biome
- `biome.json` translation files (otherwise, the biome name gets displayed in all caps by default)
- If a biome is "rare/valuable", it can be added to the list of destination biomes for Teleporting Hijinks.

Author: Mystery encounters really do be my bane huh.

Personally I think it makes more sense for the data mapping to be biome -> allowed MEs where allowed MEs

i.e.

User creates new ME -> Adds ME to appropriate array
```ts
genericMEs = [ me1, me2, me3]
civilizationMEs = [me4, me5, me6]
extreme MEs = [me7, me8]
```
Then each biome just keeps track of which MEs are allowed

```ts
biomeExample.allowedMEs = [...genericMEs, ...extremeMEs]
```

Reviewer 2: It might make more sense to do the reverse and store dis-allowed MEs, since there will be less of those.

Author: So looking at `mystery-encounter.ts` it looks like that's what the game is doing. It creates arrays of mystery encounter type's, (extremeBiomeEncounters, civilizationBiomeEncounters, anyBiomeEncounters, etc) then populates a map of BiomeId -> MEs in that biome, and then the various BiomeEncounters

### Other code that needs updating
* Don't forget to add connections to and from the biome in `biome-link.ts`
* For Pokemon with specific forms in certain biomes, `arena.getSpeciesFormIndex` can set specific forms for specific biomes

> * The move Camoflage uses the CopyBiomeTypeAttr which changes the user's type based on the current biome
> * The move Secret Power uses the SecretPowerAttr which applies a secondary effect based on the current biome
> * The move Nature Power uses the NaturePowerAttr which changes into a different attack depending on the current biome

Reviewer 2: ```suggestion
* The move Camoflage uses the [CopyBiomeTypeAttr](../moves/move-attrs/copy-biome-type-attr.ts) which changes the user's type based on the current biome
* The move Nature Power uses the [NaturePowerAttr](../moves/move-attrs/nature-power-attr.ts) which changes into a different attack depending on the current biome
* The move Secret Power uses the [SecretPowerAttr](../moves/move-attrs/secret-power-attr.ts) which applies a secondary effect based on the current biome
```
This works in VSCode, not sure if the links will work on GitHub.
